File: source/private/isCategoryReferenceIn.m
function tf = isCategoryReferenceIn(aGraph, category)
%ISCATEGORYREFERENCE    Given a graph where each node represents a variant,
%get an array that says whether each node is the category reference for the
%given category.

% Copyright 2023 Acadèmia de su Sardu APS
graphNodes = aGraph.Nodes;
numNodes = aGraph.numnodes();
tf = false(aGraph.numnodes(), 1);

for k = 1:numNodes
    currAttributes = graphNodes.Attributes{k};

    for j = 1:numel(currAttributes)
        if currAttributes(j).Category==category && currAttributes(j).IsCategoryReference
            tf(k) = true;
        end
    end
end
end

File: python/linguistic_variation_toolbox/_private/is_category_reference_in.py
import networkx as nx


def is_category_reference_in(graph: nx.Graph, category: str) -> list[bool]:
    """Return a boolean list: True if node i is the category reference for `category`."""
    result = []
    for node in graph.nodes:
        attrs = graph.nodes[node]["Attributes"]
        ref = any(
            a.category == category and a.is_category_reference for a in attrs
        )
        result.append(ref)
    return result
